File: agent/adk-agents/challenges-agent.ts
import { LlmAgent, SequentialAgent } from '@google/adk';
import * as fs from 'fs';
import * as path from 'path';
import dotenv from 'dotenv';
import { problemsCreatorInstruction, solutionsCreatorInstruction } from '../utils/prompts';
import { OpenRouterLlm } from '../utils/openrouter-llm';

dotenv.config();

const getModel = () => {
    if (process.env.CHALLENGES_MODEL) {
        return process.env.CHALLENGES_MODEL.trim();
    }
    if (process.env.RESEARCH_MODEL) {
        const models = process.env.RESEARCH_MODEL.split(',');
        return models[0].trim();
    }
    return process.env.GOOGLE_GENAI_MODEL || 'gemini-2.0-flash';
};

const modelName = getModel();
const modelInstance = new OpenRouterLlm({ model: modelName });

const problemsAgent = new LlmAgent({
    name: 'study_buddy_problems_creator',
    description:
        'Generates 3 well-explained programming challenges (1 easy, 1 medium, 1 hard) without solutions.',
    model: modelInstance,
    instruction: problemsCreatorInstruction,
    outputKey: 'problems_result',
    generateContentConfig: {
        maxOutputTokens: 6000,
        temperature: 0.3,
    },
});

const solutionsAgent = new LlmAgent({
    name: 'study_buddy_solutions_creator',
    description:
        'Generates functional solutions and code explanations for each of the generated problems.',
    model: modelInstance,
    instruction: solutionsCreatorInstruction,
    outputKey: 'code_challenges',
    generateContentConfig: {
        maxOutputTokens: 8000,
        temperature: 0.2,
    },
});

const challengesAgent = new SequentialAgent({
    name: 'study_buddy_challenges',
    description: 'Generates 3 algorithmic coding challenges and their solutions sequentially.',
    subAgents: [problemsAgent, solutionsAgent],
});

export default challengesAgent;

File: agent_python/adk_agents/challenges_agent.py
import os
from google.adk.agents import LlmAgent
from google.genai import types

from utils.prompts import challenges_creator_instruction
from utils.openrouter_llm import OpenRouterLlm

def get_model() -> str:
    challenges_model = os.getenv("CHALLENGES_MODEL")
    if challenges_model:
        return challenges_model.strip()
    research_model = os.getenv("RESEARCH_MODEL")
    if research_model:
        models = research_model.split(",")
        return models[0].strip()
    return os.getenv("GOOGLE_GENAI_MODEL", "gemini-2.0-flash")

model_name = get_model()
model_instance = OpenRouterLlm(model=model_name)

challenges_agent = LlmAgent(
    name="study_buddy_challenges",
    description="Generates 3 well-explained programming challenges (1 easy, 1 medium, 1 hard).",
    model=model_instance,
    instruction=challenges_creator_instruction,
    # Omit output_key so the agent always generates a fresh response
    generate_content_config=types.GenerateContentConfig(
        max_output_tokens=8000,
        temperature=0.3,
    ),
)
